source/agora/consensus/ValidatorSet.d

```diff
@@ -436,86 +406,88 @@ public class ValidatorSet
+            preimage = Hash(row.peek!(char[])(0));
+            distance = row.peek!ushort(1);
+            return true;
         }
         catch (Exception ex)
         {
-            log.error("Database operation error on addPreimage: {}, " ~
-                "Preimage: {}", ex.msg, preimage);
+            log.error("Error calling getOlderPreimageDistance for key {}: {}",
+                enroll_hash, ex);
             return false;
         }
+    }
 
-        // check the validity of new pre-image based on the stored pre-image
-        PreImageInfo stored_image = this.getPreimage(preimage.enroll_key);
-        if (stored_image != PreImageInfo.init)
-        {
-            if (auto reason = isInvalidReason(
-                preimage.distance, preimage.hash, stored_image.distance,
-                stored_image.hash, Enrollment.ValidatorCycle))
-            {
-                log.info("Invalid preimage data: {}, Data was: ", reason,
-                    preimage);
-                return false;
-            }
-        }
+    /***************************************************************************
 
-        // insert the pre-image into the table
-        try
+        Add a pre-image information to a validator data
+
+        Params:
+            preimage = the pre-image information to add
+
+        Returns:
+            true if the pre-image information has been added to the validator
+
+    ***************************************************************************/
+
+    public bool addPreimage (const ref PreImageInfo preimage) @trusted nothrow
+    {
+        Hash prev_preimage;
+        ushort prev_distance;
+        if (!this.getOlderPreimageDistance(preimage.enroll_key,
+            preimage.distance, prev_preimage, prev_distance))
         {
-            serializeToBuffer(preimage, buffer);
+            log.info("Rejected pre-image: Validator either not enrolled or " ~
+                "distance is outdated {}", preimage);
+            return false;
         }
-        catch (Exception ex)
+
+        if (auto reason = isInvalidReason(preimage.distance,
+            preimage.hash, prev_distance, prev_preimage,
+            Enrollment.ValidatorCycle))
         {
-            log.error("Serialization error on addPreimage: {}, Preimage: {}",
-                ex.msg, preimage);
+            log.info("Invalid pre-image data: {}. Pre-image: {}",
+                reason, preimage);
             return false;
         }
 
+        // update the preimage info
         try
         {
             () @trusted {
-                this.db.execute("UPDATE validator_set SET preimage = ? " ~
-                    "WHERE key = ?", buffer, preimage.enroll_key.toString());
+                this.db.execute("UPDATE validator_set SET preimage = ?, " ~
+                    "distance = ? WHERE key = ?",
+                    preimage.hash.toString(), preimage.distance,
+                    preimage.enroll_key.toString());
             }();
         }
```

Comment:
Instead of doing a `SELECT` then `UPDATE`, do `UPDATE` + `WHERE` ?

Reply:
Not a bad idea.

Reply:
Although, how would I verify the preimage hash matches? I still need to read the old distance and then call `isInvalidReason`.

Reply:
Damn, right.